Fix mocking

MoonBow-1 opened the pull request:
Fixes #27 
Closes #50 

- Removes PowerMock
- Uses the TestServer implementation from jla_01

MoonBow-1 committed:
Add private method for creating HeaderInfo in HECServiceImpl

- Uncomplicates the sendEvents method

Fix error after rebasing

Fix errors after rebasing

Remove unused vars from ServiceAndHECBatchIT after rebase

Apply Spotless after rebase

Rename variables to new object names

- Some vars were not renamed when refactoring objects, so fix the HttpEvent... references

Rename Event to EventMessage for more clarity

Move test class to correct package

Add JavaDoc to describe the Event interface

Remove assertions and thread lifetime from Jmeter test

- Causes problems if the test is not run in 30 seconds

Implement unit tests for HECTimeImplWithFallback.instant()

Add equals test for HECTimeImplWithFallback

Implement source() method for HECTimeImplWithFallback and add tests for that

Implement parsed() method for HECTimeImplWithFallback and add tests for that

Remove unused objects

Remove unused method found in JsonEvent

Use a different way of checking HECBatchTest.asHttpEventDataListTest

Implement equals and hashCode tio HeaderInfo

Method overload HECRecord.toSyslogMessage so it can accept a defaultValue

- Allows for tests to declare an expected value
- Tested by running the test file 1000 times

Cleanup variables in EventDataToSyslogMessageTest

Apply quick fixes to for-loops and variable assignments in EventDataToSyslogMessageTest

Rename ConverterTests.java to EventDataToSyslogMessageTest.java

Implement HECTime classes

- Old Time classes have been deleted
- Tests are missing and broken at the moment

Rename HttpEventData to HECRecord

Rename EventBatch to HECBatch

Implement Time objects

- Cleans up the EventTime.asTime() if-else tree

Update cfe-16-test-plan.jmx

Refactor HEC service and the batches

Rename sender variables to connections

Rename TimestampedHttpEventData to HttpEventDataImpl

Rename AckManager to Acknowledgements

Rename EventManager to EventBatch

Stop parsing events if an error occurs in JSON processing

Change IllegalStateExceptions to UnsupportedOperationExceptions for all stub objects

Fix order of keywords in test files

- "final static" -> "static final"

Make TimeObjectStub.java throw errors if hashCode or equals is called

Make TimestampedHttpEventDataStub.java throw errors if hashCode or equals is called

Make EventStub.java throw errors if hashCode or equals is called

Remove EventString.java and use Event.java as an interface for event messages

Use TimeObject for handling timestamps from JsonNodes

- Stub used when a timestamp was not found

Rename ValidatedJsonEvent.java to JsonEventImpl.java

Make ValidatedJsonEvent class final

Fix previousEvent assignment position

Remove EpochTimeString.java, since it wasn't used anymore

Remove wrapper classes and introduce Stubable

Fix wrong logic in hashCode methods with multiple fields

Add new metadataSDParam to indicate if event is generated

Make EventTime tests more readable/understandable

Accept timestamps shorter and longer than 13 characters

- Tests have been fixed

Remove empty test file

Initial change to EventTime parsing

- Tests fail at the moment

Use new field for checking if TimestampedHttpEventData is default

Do not default to null eventData in EventManager.convertData

- Additional methods and tests

Change getters to proper methods

Move functionality from private method to public one in EventManager

Update formatting in EventManager

Equals tests for EventTime and TimestampedHttpEventData

Initial EventTime implementation

Apply Spotless

Default ackID to null, as it was in the previous implementation

Tests for DefaultHttpEventData

Re-implement ackID and make DefaultHttpEventData immutable

Tests for EpochTimeString

- Adds equality methods

Add new request type to cfe-16-test-plan.jmx ti test message nodes

- Test plan runs 2000 times with 8 threads with 6 requests, totaling 96000 requests

JsonEvent and EventString classes and tests

Initial implementation for JsonEvent and EventString

Use ctor for DefaultHttpEventData instead of setter for Event in EventManager

Fix pom.xml formatting to be based on main branch

Remove dashes from the license header in log4j2.xml

Update Apache Rat license header to correct year

Remove commented out code from ServiceAndEventManagerIT

Fix ServiceAndEventManagerIT test where an ackID is not supposed to be returned

Exclude eclipse-java-formatter.xml from Apache Rat

Add Spotless beck to pom.xml after rebase

Revert license modified by accident in SyslogMessageSender

- License gotten from fa286024

Specify compiler plugin to use the '-parameters' arg

- Otherwise does compile, but doesn't start properly

Restore missing synchronized keyword removed in 063bc3a7

Use literal ints for expected loop iterations in tests

Update assertion failure messages to indicate reason, not the goal

Update JavaDoc about EpochTimeString.asEpochMillis

Add assertions for count of assertions in loops

Use assertDoesNotThrow instead of try-catch for test

Remove commented out old tests

Reformat pom.xml to have important info at the top

Revert all cosmetic changes to formatting

Update Apache RAT pattern for license

Change headers to say when copyright begins

Make HECServiceImpl.sendEvents synchronized

- Fixes tests
- Performance impact

Fix ConfigurationIT test

Add RELP server from jla_01 for usage in tests

Update Copyright to year 2025

Fix variable initialization order when parsing time as Double

Remove unused field ackID from DefaultHttpEventData

Check for time length when asking for timeSource and timeParsed

Remove PowerMock from pom.xml

Generate tests for TimestampedHttpEventData

- 4 failing assertions found in this commit

Create decorator for DefaultHttpEventData

- Time to migrate the time related tests from ServiceAndEventManagerIT to TimestampedHttpEventDataTest

Rename HttpEventData.java to DefaultHttpEventData.java

eemhu commented on pom.xml:
```diff
@@ -1,323 +1,367 @@
     </repository>
   </distributionManagement>
+  <groupId>com.teragrep</groupId>
+  <modelVersion>4.0.0</modelVersion>
```
the structure of the pom.xml file seems a bit upside down - usually the file starts with the name and properties.

MoonBow-1 replied:
The IDE reformatting/cleanup tool seems to prefer alphabetical ordering, but I'll format it to include the more important information at the beginning

MoonBow-1 replied:
Formatting has been changed in commit 2239683. Would that be better?

eemhu commented on src/main/java/com/teragrep/cfe_16/bo/EpochTimeString.java:
```diff
@@ -0,0 +1,75 @@
+    }
+
+    /**
+     * Converts the given time stamp into epoch milliseconds. Takes a HttpEventData object as a
```
comment is no longer valid

MoonBow-1 replied:
Fixed in commit 8345a9b4

eemhu commented on src/main/java/com/teragrep/cfe_16/bo/DefaultHttpEventData.java:
```diff
@@ -0,0 +1,100 @@
+        this.authenticationToken = authenticationToken;
+    }
+
+    public String getEvent() {
```
new issue for refactoring

MoonBow-1 replied:
Issue #17 created for this

eemhu commented on src/main/java/com/teragrep/cfe_16/bo/HttpEventData.java:
```diff
@@ -46,87 +46,11 @@
-    public Integer getAckID() {
-        return ackID;
-    }
+    String getEvent();
```
remove `get` prefix (or leave for another issue)

MoonBow-1 replied:
Issue #17 created for this

eemhu commented on src/main/java/com/teragrep/cfe_16/bo/TimestampedHttpEventData.java:
```diff
@@ -0,0 +1,227 @@
+    private long removeDecimal(double doubleValue) {
+        BigDecimal doubleValueWithDecimal = BigDecimal.valueOf(doubleValue);
+        String stringValue = doubleValueWithDecimal.toString();
+        String stringValueWithoutDecimal = stringValue.replace(".", "");
```
any reason to not use `Double.valueOf(...).longValue()` instead?

MoonBow-1 replied:
`Double.valueOf(...).longValue()`'s description:
> Returns the value of this Double as a long after a narrowing primitive conversion.

I tested this method, and the resulting long was missing 3 digits compared to the current BigDecimal implementation.

If I didn't understand the desired implementation, please ask for further implementation from my part.

eemhu replied:
Okay after reading the code block again, I can see that the decimal point is just removed but the decimals are left appended in the resulting number.

eemhu commented on src/test/java/com/teragrep/cfe_16/ConverterTests.java:
```diff
@@ -287,15 +305,16 @@ public void SDElementsTest() {
         for (int i = 0; i < returnedSDParams.size(); i++) {
-            assertTrue("SDParam '" + returnedSDParams.get(i) + "' should not be in returned SDElement.",
-                    supposedSDParams.contains(returnedSDParams.get(i)));
+            assertTrue(
```
loops executed should be asserted

MoonBow-1 replied:
Fixed in commit f7b6142b

eemhu commented on src/test/java/com/teragrep/cfe_16/bo/TimestampedHttpEventDataTest.java:
```diff
@@ -0,0 +1,498 @@
+        JsonNode jsonNode;
+
+        try {
+            jsonNode = objectMapper.readTree(content);
```
should use `Assertions.assertDoesNotThrow()` instead of try-catch

MoonBow-1 replied:
Fixed in commit 8235059e

eemhu commented on src/test/java/com/teragrep/cfe_16/bo/TimestampedHttpEventDataTest.java:
```diff
@@ -0,0 +1,498 @@
+    }
+
+
+    /*
```
should remove if not relevant

MoonBow-1 replied:
Fixed in commit e8eea13b

eemhu commented on src/test/java/com/teragrep/cfe_16/it/SendMultipleEventsIT.java:
```diff
@@ -0,0 +1,158 @@
+        for (Future<String> f : futures) {
+            final String actualResponse = f.get();
+            Assertions.assertTrue(supposedResponses.contains(actualResponse),
+                "Service should return JSON object with fields 'text', 'code' and 'ackID' (ackID "
```
assert loops executed

MoonBow-1 replied:
Fixed in commit f7b6142b

kortemik commented on pom.xml:
```diff
@@ -1,319 +1,374 @@
     <dependency>
-      <groupId>systems.manifold</groupId>
       <artifactId>manifold-all</artifactId>
+      <groupId>systems.manifold</groupId>
```
is manifold needed?

MoonBow-1 replied:
Apparently not, project compiles and verifies without the plugin.
Should a new issue be made for the removal?

kortemik replied:
yes please

kortemik commented on src/main/java/com/teragrep/cfe_16/ThirdParty/SyslogMessageSender/SyslogMessageSender.java:
```diff
@@ -1,17 +1,47 @@
 /*
```
license must not change

MoonBow-1 replied:
License reverted in c7b94908

kortemik commented on src/main/resources/log4j2.xml:
```diff
@@ -1,20 +1,66 @@
 <?xml version="1.0" encoding="UTF-8"?>
+<!--
+  ~ HTTP Event Capture to RFC5424 CFE_16
```
please remove dashes at line start

MoonBow-1 replied:
Fixed in commit 11148e0e

kortemik commented:
please rebase

MoonBow-1 commented:
> please rebase

Rebase done

MoonBow-1 commented:
@kortemik Rebase done.
#24 created, since there are some files that require formatting to be done.

kortemik commented on src/main/java/com/teragrep/cfe_16/Converter.java:
```diff
@@ -150,11 +149,6 @@ private void setStructuredDataParams(HttpEventData eventData) {
 
-        if (eventData.getAckID() != null) {
-            LOGGER.debug("Setting ack id");
-            metadataSDE.addSDParam("ack_id", String.valueOf(eventData.getAckID()));
```
why ack_id is being removed?

MoonBow-1 replied:
ackID re-implemented in 460bc8e7

kortemik commented on src/main/java/com/teragrep/cfe_16/EventManager.java:
```diff
@@ -220,111 +218,27 @@ private HttpEventData verifyJsonData(String eventInJson, HttpEventData previousE
-            else {
+                final String eventString = event.toString();
+                DefaultHttpEventData defaultHttpEventData = new DefaultHttpEventData();
+                defaultHttpEventData.setEvent(eventString);
```
why setter instead of ctor arg?

kortemik commented on src/main/java/com/teragrep/cfe_16/EventManager.java:
```diff
@@ -142,14 +143,14 @@ public ObjectNode convertData(
             String jsonObjectStr = parser.next().toString();
             eventData = verifyJsonData(jsonObjectStr, previousEvent);
-            eventData = assignMetaData(eventData, authToken, channel);
+            assignMetaData(eventData, authToken, channel);
```
method is using setters to modify method call args, which is horrible. please remove the method and move at least the calling code here.

kortemik commented on src/main/java/com/teragrep/cfe_16/bo/EpochTimeString.java:
```diff
@@ -0,0 +1,73 @@
+
+    /**
+     * Converts the given time stamp into epoch milliseconds. If the time value
+     * in the object has 13 digits, it means that time has been already given in epoch milliseconds.
```
why guesswork is required?

MoonBow-1 replied:
I asked the original writer about the origin of this, but there wasn't a conclusive answer.

Should we trust that the incoming timestamp is in correct format?

kortemik replied:
check first if numeric only, if numeric only, then it is epoch, otherwise it might be known timestamp, try conversion to the known format and if it fails generate one for the time of the receive?

MoonBow-1 replied:
Should we try to convert a possible String timestamp from predetermined formats to a long timestamp? This was mentioned by @elliVM in our discussion, and formats are used in [PTH_10](https://github.com/teragrep/pth_10/blob/7c78cf354cf462de22321bd6485c2ad8a888ce69/src/main/java/com/teragrep/pth10/ast/DefaultTimeFormat.java#L62)

kortemik reviewed (changes requested):
please see above comments and reply. I will check tests after the /main/ is resolved

kortemik commented on src/main/java/com/teragrep/cfe_16/EventManager.java:
```diff
@@ -142,30 +146,46 @@ public ObjectNode convertData(
          * variable.
          */
-        HttpEventData eventData = null;
+        TimestampedHttpEventData eventData = new TimestampedHttpEventData();
```
why default values are being used instead of a stub that throws when it is attempted to use? datapath must not contain bogus values such as defaults are

MoonBow-1 replied:
Fixed in commit f759a236

kortemik commented on src/main/java/com/teragrep/cfe_16/bo/DefaultHttpEventData.java:
```diff
@@ -0,0 +1,113 @@
+    private final String authenticationToken;
+    private final Integer ackID;
+
+    public DefaultHttpEventData() {
```
this must be a stub

MoonBow-1 replied:
Fixed in commit f759a236

kortemik commented on src/main/java/com/teragrep/cfe_16/bo/DefaultHttpEventData.java:
```diff
@@ -0,0 +1,113 @@
+    }
+
+    public DefaultHttpEventData(String event) {
+        this("", event, "", null);
```
this must be a stub

MoonBow-1 replied:
Fixed in commit f759a236

kortemik commented on src/main/java/com/teragrep/cfe_16/bo/DefaultHttpEventData.java:
```diff
@@ -0,0 +1,113 @@
+    }
+
+    @Override
+    public int hashCode() {
```
hashCode implementation is not according to standards see Procedure: Object Equality section 3

MoonBow-1 replied:
Fixed in commit 27e76ccb

kortemik commented on src/main/java/com/teragrep/cfe_16/bo/XForwardedForImpl.java:
```diff
@@ -0,0 +1,89 @@
+    }
+
+    @Override
+    public int hashCode() {
```
hashCode implementation is not according to standards see Procedure: Object Equality section 3

MoonBow-1 replied:
Fixed in commit 27e76ccb

kortemik commented on src/main/java/com/teragrep/cfe_16/bo/DefaultHttpEventData.java:
```diff
@@ -0,0 +1,113 @@
+
+import java.util.Objects;
+
+public final class DefaultHttpEventData implements HttpEventData {
```
please describe the responsibility of this object

MoonBow-1 replied:
After thinking about the responsibilities, this object was just a wrapper for the data underneath.
Class removed/renamed in commit f759a236, where a stub object was taken into use

kortemik commented on src/main/java/com/teragrep/cfe_16/bo/EpochTimeString.java:
```diff
@@ -0,0 +1,89 @@
+    }
+
+    /**
+     * Converts the given time stamp into epoch milliseconds. If the time value in the object has 13 digits, it means
```
please reason the logic in the javadoc. current explanation and the code do not match

MoonBow-1 replied:
Since we [discussed](https://github.com/teragrep/cfe_16/issues/27#issuecomment-2916044048) the responsibility of the logger, where we will just accept any timestamp, EpochTimeString was removed in commit fc11896f, since it wasn't used anymore.
So CFE_16 will accept epoch timestamps shorter and longer than 13 characters.

kortemik commented on src/main/java/com/teragrep/cfe_16/bo/TimestampedHttpEventData.java:
```diff
@@ -0,0 +1,168 @@
+    private final long timeAsLong;
+    private final boolean timeParsed;
+    private final boolean isDefault;
+
```
please arrange the main constructor as last

MoonBow-1 replied:
Fixed in commit f759a236, since other constructors were removed

kortemik commented on src/main/java/com/teragrep/cfe_16/bo/TimestampedHttpEventData.java:
```diff
@@ -0,0 +1,168 @@
+        this(defaultHttpEventData, "timeSource", "time", 0L, false, isDefault);
+    }
+
+    public TimestampedHttpEventData() {
```
default value is a stub and must be implementad as such

MoonBow-1 replied:
Fixed in commit f759a236

kortemik commented on src/main/java/com/teragrep/cfe_16/bo/TimestampedHttpEventData.java:
```diff
@@ -0,0 +1,168 @@
+        this(defaultHttpEventData, false);
+    }
+
+    public TimestampedHttpEventData(final DefaultHttpEventData defaultHttpEventData, boolean isDefault) {
```
default is a stub and must be implemented as such

MoonBow-1 replied:
Fixed in commit f759a236

kortemik commented on src/main/java/com/teragrep/cfe_16/bo/TimestampedHttpEventData.java:
```diff
@@ -0,0 +1,168 @@
+        this(eventData, timeSource, time, timeAsLong, timeParsed, false);
+    }
+
+    public TimestampedHttpEventData(final DefaultHttpEventData defaultHttpEventData) {
```
default is a stub and must be implemented as such

MoonBow-1 replied:
Fixed in commit f759a236

kortemik commented on src/main/java/com/teragrep/cfe_16/bo/TimestampedHttpEventData.java:
```diff
@@ -0,0 +1,168 @@
+    }
+
+    public String timeSource() {
+        if (this.time == null) {
```
please use a stub time object instead of null values within the object

kortemik commented on src/main/java/com/teragrep/cfe_16/bo/TimestampedHttpEventData.java:
```diff
@@ -0,0 +1,168 @@
+    }
+
+    public boolean timeParsed() {
+        if (this.time == null) {
```
please use a stub time object instead of null values within the object

kortemik commented on src/main/java/com/teragrep/cfe_16/bo/TimestampedHttpEventData.java:
```diff
@@ -0,0 +1,168 @@
+    }
+
+    @Override
+    public int hashCode() {
```
hashCode implementation is not according to standards see Procedure: Object Equality section 3

MoonBow-1 replied:
Fixed in commit 27e76ccb

kortemik commented on src/main/java/com/teragrep/cfe_16/event/DefaultJsonEvent.java:
```diff
@@ -0,0 +1,88 @@
+import com.fasterxml.jackson.databind.JsonNode;
+import java.util.Objects;
+
+public final class DefaultJsonEvent implements JsonEvent {
```
i assume this is a hidden json parser object. it should rather be a JsonEventImpl and return safe types, such as TimeStub and EventStub in case they are null so that null values are not used. Same goes for other types so that they must represent their true form which may be for the node for example "Record" or "Payload" and have just a responsibility to be a byte array or String

kortemik commented on src/main/java/com/teragrep/cfe_16/response/JsonResponse.java:
```diff
@@ -0,0 +1,87 @@
+    }
+
+    @Override
+    public int hashCode() {
```
hashCode implementation is not according to standards see Procedure: Object Equality section 3

MoonBow-1 replied:
Fixed in commit 27e76ccb

kortemik commented on src/main/java/com/teragrep/cfe_16/event/EventString.java:
```diff
@@ -0,0 +1,87 @@
+import com.fasterxml.jackson.databind.ObjectMapper;
+import java.util.Objects;
+
+public final class EventString {
```
please describe the responsibility of this Object? Is it just a json parser that returns the event list as a node? perhaps it should be return the Event (Record) as a collection instead of a JsonNode?

kortemik commented on src/main/java/com/teragrep/cfe_16/event/EventTime.java:
```diff
@@ -0,0 +1,178 @@
+import java.math.BigDecimal;
+import java.util.Objects;
+
+public final class EventTime {
```
please describe the responsibility of this object. is this an previousEvent to nextEvent time transferer? in such case it might make sense to explore the time aspect of the problem itself instead of implementing a transferer object

kortemik commented on src/main/java/com/teragrep/cfe_16/event/EventTime.java:
```diff
@@ -0,0 +1,178 @@
+        final boolean timeParsed;
+
+        // No time provided in the event
+        if (timeObject == null) {
```
stub instead of a null is the only permissible approach

kortemik commented on src/main/java/com/teragrep/cfe_16/event/EventTime.java:
```diff
@@ -0,0 +1,178 @@
+        // No time provided in the event
+        if (timeObject == null) {
+            // Previous event does not have a proper time
+            if (previousEvent == null || previousEvent.isDefault()) {
```
stub instead of a null is the only permissible approach

kortemik commented on src/main/java/com/teragrep/cfe_16/event/EventTime.java:
```diff
@@ -0,0 +1,178 @@
+    }
+
+    @Override
+    public int hashCode() {
```
hashCode implementation is not according to standards see Procedure: Object Equality section 3

MoonBow-1 replied:
Fixed in commit 27e76ccb

kortemik commented on src/main/java/com/teragrep/cfe_16/event/JsonEvent.java:
```diff
@@ -43,13 +43,15 @@
+import com.fasterxml.jackson.databind.JsonNode;
 
-public class Cfe16ApplicationTests {
+public interface JsonEvent {
```
JsonEvent is a good approach. the methods should be rather replaced with asEvent (asRecord preferred) and the Event object itself should contain the values. Responsibility of the JsonEvent would be then just to provide a view into the content as a parsed object.

MoonBow-1 replied:
Method has been replaced with asEvent(),  which will return a `EventMessage`, which can be a stub or a proper implementation.

EventMessage interface:
```java
/**
 * Interface for objects that work with the Event (message)
 */
public interface EventMessage extends Stubable {
    String asString();
}
```

kortemik commented on src/main/java/com/teragrep/cfe_16/event/ValidatedJsonEvent.java:
```diff
@@ -0,0 +1,116 @@
+import com.teragrep.cfe_16.exceptionhandling.EventFieldMissingException;
+import java.util.Objects;
+
+public class ValidatedJsonEvent implements JsonEvent {
```
final missing. please reconsider implementing methods as commented on the interface itself

kortemik commented on src/main/java/com/teragrep/cfe_16/event/ValidatedJsonEvent.java:
```diff
@@ -0,0 +1,116 @@
+    }
+
+    /**
+     * Return the time from the {@link #jsonEvent}. If it is null, it is the responsibility of someone else to generate
```
null values are not permissible, please use stub instead

kortemik commented on src/test/java/com/teragrep/cfe_16/it/ServiceAndEventManagerIT.java:
```diff
@@ -334,9 +307,8 @@ public void getAcksWithUnusedChannel() {
-        /*AckManager ackManager = new AckManager();*/
         String allEventsInJson = "{\"sourcetype\": \"mysourcetype\", \"event\": \"Hello, world!\", \"host\": \"localhost\", \"source\": \"mysource\", \"index\": \"myindex\"}";
-        String supposedResponse = "{\"text\":\"Success\",\"code\":0,\"ackID\":2}";
+        String supposedResponse = "{\"text\":\"Success\",\"code\":0,\"ackID\":0}";
```
why expected value changes?

MoonBow-1 replied:
In the previous implementation, (I assume, since ackManager wasn't touched) the server was not reset after each test, but the socket was, so the ackID was incremented before the test was run, resulting in an ackID of NOT a zero. Current implementation will reset the ackID after each test. Current implementation tests the incrementing ackID if the ackManager isn't reset in between the tests in https://github.com/teragrep/cfe_16/pull/15/files/1b5ec29c7a3837a2e84b61d33e04f7a60d855827#diff-ff882e40edea750c05377048bd8a3ec0a3f55db983db9a50e683bd1db07069a4R203-R213 (sendEventsAndGetAcksTest)

kortemik commented on src/test/java/com/teragrep/cfe_16/it/ServiceAndEventManagerIT.java:
```diff
@@ -566,10 +378,10 @@ public void sendingMultipleEventsTest() {
     public void sendingMultipleEventsWithDefaultChannelTest() {
-        AckManager ackManager = new AckManager();
         String allEventsInJson = "{\"event\": \"Pony 1 has left the barn\", \"sourcetype\": \"mysourcetype\", \"time\": 1426279439}{\"event\": \"Pony 2 has left the barn\"}{\"event\": \"Pony 3 has left the barn\", \"sourcetype\": \"newsourcetype\"}{\"event\": \"Pony 4 has left the barn\"}";
-        String supposedResponse = "{\"text\":\"Success\",\"code\":0,\"ackID\":0}";
```
why expected value changes?

MoonBow-1 replied:
In the previous implementation, (I assume, since ackManager wasn't touched) the server was not reset after each test, but the socket was, so the ackID was incremented before the test was run, resulting in an ackID of NOT a zero. Current implementation will reset the ackID after each test. Current implementation tests the incrementing ackID if the ackManager isn't reset in between the tests in https://github.com/teragrep/cfe_16/pull/15/files/1b5ec29c7a3837a2e84b61d33e04f7a60d855827#diff-ff882e40edea750c05377048bd8a3ec0a3f55db983db9a50e683bd1db07069a4R203-R213 (sendEventsAndGetAcksTest)

elliVM commented on src/main/java/com/teragrep/cfe_16/bo/XForwardedForStub.java:
```diff
@@ -0,0 +1,69 @@
+        return true;
+    }
+
+    @Override
```
I don't think this stub object really needs a working implementation of `hashCode` and `equals` methods,  maybe they could throw `IllegalStateException` instead?

MoonBow-1 replied:
Made EventStub throw an IllegalStateException in commit e0d26b0f, test added there as well

elliVM commented on src/test/java/com/teragrep/cfe_16/event/EventImplTest.java:
```diff
@@ -45,52 +45,43 @@
 
     @Test
     @DisplayName("Happy equals test")
     void happyEqualsTest() {
```
Perhaps we should test `equals` and `hashCode` methods with [EqualsVerifier](https://mvnrepository.com/artifact/nl.jqno.equalsverifier/equalsverifier) as instructed in the Object Equality procedure.

MoonBow-1 replied:
Created a new issue for that: #29

elliVM commented on src/test/java/com/teragrep/cfe_16/event/EventStubTest.java:
```diff
@@ -0,0 +1,71 @@
+import org.junit.jupiter.api.DisplayName;
+import org.junit.jupiter.api.Test;
+
+class EventStubTest {
```
Add tests for `equals` and `hashCode` methods if they are kept

MoonBow-1 replied:
Made EventStub throw an Exception in commit e0d26b0f, test added there as well

elliVM commented on src/test/java/com/teragrep/cfe_16/event/JsonEventImplMessagesTest.java:
```diff
@@ -47,59 +47,55 @@
+        final Event returnedEvent = jsonEventImpl.asEvent();
 
-        Assertions.assertEquals("jsonEvent node's event not valid", exception.getMessage());
+        Assertions.assertEquals(new EventStub(), returnedEvent);
```
Here we could perhaps better assert for stubness with `Assertions.assertTrue(returnedEvent.isStub())`.
I think the current assertion relies on the implementation of equals method.
Or keep both

MoonBow-1 replied:
Fixed in commit e0d26b0f, since hashCode() and equals() now throw an Exception, indicating that they shouldn't be used for the Stub:

```java
Assertions.assertTrue(returnedEvent::isStub);
```

elliVM commented on src/test/java/com/teragrep/cfe_16/event/JsonEventImplMessagesTest.java:
```diff
@@ -47,59 +47,55 @@
+        final Event returnedEvent = jsonEventImpl.asEvent();
 
-        Assertions.assertEquals("jsonEvent node's event not valid", exception.getMessage());
+        Assertions.assertEquals(new EventStub(), returnedEvent);
```
Same as above we could add  `Assertions.assertTrue(returnedEvent.isStub())`

MoonBow-1 replied:
Fixed in commit e0d26b0f, since hashCode() and equals() now throw an Exception, indicating that they shouldn't be used for the Stub:

```java
Assertions.assertTrue(returnedEvent::isStub);
```

elliVM commented on src/test/java/com/teragrep/cfe_16/event/JsonEventImplTest.java:
```diff
@@ -47,69 +47,64 @@
+        final Event returnedEvent = jsonEventImpl.asEvent();
 
-        Assertions.assertEquals("event field is missing", exception.getMessage());
+        Assertions.assertEquals(new EventStub(), returnedEvent);
```
`Assertions.assertTrue(returnedEvent.isStub())` here also

MoonBow-1 replied:
Fixed in commit e0d26b0f, since hashCode() and equals() now throw an Exception, indicating that they shouldn't be used for the Stub:

```java
Assertions.assertTrue(returnedEvent::isStub);
```

elliVM commented on src/test/java/com/teragrep/cfe_16/event/JsonEventImplTest.java:
```diff
@@ -47,69 +47,64 @@
+        final Event returnedEvent = jsonEventImpl.asEvent();
 
-        Assertions.assertEquals("jsonEvent node's event not valid", exception.getMessage());
+        Assertions.assertEquals(new EventStub(), returnedEvent);
```
`Assertions.assertTrue(returnedEvent.isStub())` here also

MoonBow-1 replied:
Fixed in commit e0d26b0f, since hashCode() and equals() now throw an Exception, indicating that they shouldn't be used for the Stub:

```java
Assertions.assertTrue(returnedEvent::isStub);
```

elliVM commented on src/test/java/com/teragrep/cfe_16/event/JsonEventImplTest.java:
```diff
@@ -47,69 +47,64 @@
+        final Event returnedEvent = jsonEventImpl.asEvent();
 
-        Assertions.assertEquals("jsonEvent node's event not valid", exception.getMessage());
+        Assertions.assertEquals(new EventStub(), returnedEvent);
```
`Assertions.assertTrue(returnedEvent.isStub())` here also

MoonBow-1 replied:
Fixed in commit e0d26b0f, since hashCode() and equals() now throw an Exception, indicating that they shouldn't be used for the Stub:

```java
Assertions.assertTrue(returnedEvent::isStub);
```

elliVM reviewed (changes requested):
New interfaces and stub objects looks good to me. I left some comments about the `equals` and `hashCode` and their testing

elliVM reviewed (dismissed):
requested changes implemented for  commit `09ebfa2` which I looked at

kortemik commented on src/main/java/com/teragrep/cfe_16/EventManager.java:
```diff
@@ -142,30 +144,41 @@ public ObjectNode convertData(
+                previousEvent = eventData;
+            }
+            catch (JsonProcessingException e) {
+                LOGGER.error("Problem processing JsonObjectString <{}>", jsonObjectStr);
```
im this case, the json object should be passed through as-is as it was broken and the program couldn't do anything with it. continue skips the whole thing.

kortemik commented on src/main/java/com/teragrep/cfe_16/bo/TimestampedHttpEventDataStub.java:
```diff
@@ -0,0 +1,106 @@
+
+    @Override
+    public Event event() {
+        throw new IllegalStateException("TimestampedHttpEventDataStub does not support this");
```
unsupported operation exception is better

kortemik commented on src/main/java/com/teragrep/cfe_16/event/EventTime.java:
```diff
@@ -0,0 +1,198 @@
+    private final HttpEventData previousEvent;
+    private final TimeObject timeObject;
+
+    public EventTime(
```
why Time has so many responsibilities, events have Time but Time has no channel, event, token, ack nor data. If this is something else, then the naming is off.

kortemik commented on src/main/java/com/teragrep/cfe_16/event/TimeObjectStub.java:
```diff
@@ -0,0 +1,99 @@
+
+    @Override
+    public boolean isDouble() {
+        throw new IllegalStateException("TimeObjectStub does not support this");
```
UnsupportedOperationException

kortemik commented:
@MoonBow-1 please rebase

MoonBow-1 committed:
Time objects from the branch and all tests now pass

- Jmeter's 1st run fails, but subsequent ones pass, requires investigating

MoonBow-1 committed:
Fix code quality for time package

MoonBow-1 committed:
Finalize variables for HECRecord files

MoonBow-1 committed:
Remove modifications to SyslogMessageSender and exclude it from future formatting changes

MoonBow-1 committed:
Change Exception messages for EventMessageStub

MoonBow-1 committed:
Fix code quality issues for HECBatch and SyslogBatch

MoonBow-1 committed:
Clarify Javadoc for HECBatch.toHECRecordList

eemhu commented on src/main/java/com/teragrep/cfe_16/bo/HECRecordStub.java:
```diff
@@ -45,89 +45,60 @@
-    public void setAckID(Integer ackID) {
-        this.ackID = ackID;
+    @Override
+    public int hashCode() {
```
isStub as field variable, implement equals and hashcode

MoonBow-1 replied:
Field added and methods implemented in commit 71cdfd8f

eemhu replied:
https://github.com/teragrep/cfe_16/commit/71cdfd8ff0a5dbd03373ab11a752e89855968af8#r177620492

eemhu commented on src/main/java/com/teragrep/cfe_16/event/time/HECTime.java:
```diff
@@ -0,0 +1,57 @@
+
+    long instant(final long defaultValue);
+
+    boolean parsed();
```
perhaps rename to` isParsed() `as the method returns boolean. `parsed() `would fit better if the return value was some sort of ParsedTime.

MoonBow-1 replied:
Method renamed in commit 95a636c0

eemhu commented on src/main/java/com/teragrep/cfe_16/event/time/HECTimeImpl.java:
```diff
@@ -0,0 +1,245 @@
+
+public final class HECTimeImpl implements HECTime {
+
+    private final JsonEvent timeNode1;
```
-> time(Json)Event?

MoonBow-1 replied:
Field renamed in commit dc6d7981

eemhu commented on src/main/java/com/teragrep/cfe_16/event/time/HECTimeImpl.java:
```diff
@@ -0,0 +1,245 @@
+                timeNode = new TextNode("");
+            }
+        }
+        catch (final EventFieldException e) {
```
is this exception something worth logging?

MoonBow-1 replied:
Exception logged and logic changed a bit in commit 56ec81e1

- Does not check if the `jsonEvent` has time before requesting it, allowing the catch to actually work

Should the JSON payload be logged, or just the Java class signature?

eemhu replied:
Not sure if logging the payload is necessarily a good idea, perhaps if there is some id that could be used?

MoonBow-1 replied:
UUID is used when the actual parsing fails, https://github.com/MoonBow-1/cfe_16/blob/fix-mocking/src/main/java/com/teragrep/cfe_16/response/ExceptionEvent.java#L70, but this try-catch will not stop the parsing, so not sure if we can get an ID from here that's meaningful

eemhu replied:
Perhaps it could be left for another issue to see about adding the UUID in that case?

MoonBow-1 replied:
Issue created: #53 
Removed the logging of the JSON payload in commit e92cd244

eemhu commented on src/main/java/com/teragrep/cfe_16/event/time/HECTimeImpl.java:
```diff
@@ -0,0 +1,245 @@
+
+    @Override
+    public String source() {
+        final String returnedSource;
```
could these similar codeblocks be combined in some way?

MoonBow-1 replied:
Methods somewhat combined in commit 342997e0

- `source()` relies on the return value of `isParsed()`, since their values are connected

eemhu commented on src/main/java/com/teragrep/cfe_16/event/time/HECTimeStub.java:
```diff
@@ -0,0 +1,69 @@
+    }
+
+    @Override
+    public boolean isStub() {
```
equals&hashCode can be added if isStub is set as a field variable.

MoonBow-1 replied:
Fixed in commit 88e4126e

eemhu replied:
use Objects.hashCode for hashCode function.

MoonBow-1 replied:
Replaced Boolean.hashCode() with Objects.hashCode() in commit fe722c7d

eemhu commented on src/main/java/com/teragrep/cfe_16/event/EventMessageStub.java:
```diff
@@ -0,0 +1,69 @@
+    public int hashCode() {
+        throw new UnsupportedOperationException("EventMessageStub does not support this");
+    }
+
```
equals & hashCode

MoonBow-1 replied:
Fixed in commit b2eeba88

eemhu commented on src/main/java/com/teragrep/cfe_16/event/JsonEventImpl.java:
```diff
@@ -102,10 +102,27 @@ public JsonNode asPayloadJsonNode() {
+
+    @Override
+    public boolean hasTime() {
+        final boolean hasTime;
```
`return jsonNode.has("time")`

MoonBow-1 replied:
Simplified in commit d5301b02

eemhu commented on src/main/java/com/teragrep/cfe_16/service/HECServiceImpl.java:
```diff
@@ -96,21 +104,37 @@ public class HECServiceImpl implements HECService {
+        try {
+            this.connection = ConnectionFactory
+                    .createConnection(
+                            this.configuration.getSysLogProtocol(), this.configuration.getSyslogHost(),
```
SysLog->Syslog

MoonBow-1 replied:
SysLog renamed to Syslog in commit d4331752

eemhu commented on src/main/java/com/teragrep/cfe_16/service/HECServiceImpl.java:
```diff
@@ -96,21 +104,37 @@ public class HECServiceImpl implements HECService {
+    }
+
     @Override
     // @LogAnnotation(type = LogType.METRIC_COUNTER)
```
Remove commented-out code

MoonBow-1 replied:
Commented out code removed in commit 9ddece4b

eemhu commented on src/main/java/com/teragrep/cfe_16/service/HECServiceImpl.java:
```diff
@@ -248,13 +269,48 @@ public JsonNode getAcks(HttpServletRequest request, String channel, JsonNode req
         return new ResponseEntity<String>("HEC is available and accepting input", HttpStatus.OK);
     }
+
+    private HeaderInfo headerInfo(final HttpServletRequest request) {
```
HeaderInfo object ?

MoonBow-1 replied:
Can I get some clarification on this question?

eemhu replied:
Could the HeaderInfo object include this functionality? To convert HttpServletRequest to HeaderInfo

MoonBow-1 replied:
Should the HeaderInfo take `HttpServletRequest` as a constructor parameter, and HeaderInfo would parse the information from there, instead of having `HECServiceImpl` parse the information required for HeaderInfo?

As in before:
```java
public HeaderInfo(
            final XForwardedFor xForwardedFor,
            final XForwardedHost xForwardedHost,
            final XForwardedProto xForwardedProto
    ) {
...
```

To:
```java
public HeaderInfo(final HttpServletRequest httpServletRequest) {
...
```

eemhu replied:
Yes I think that would make sense

MoonBow-1 replied:
Implementation done in commit 9d3129a0

MoonBow-1 committed:
Make isStub a field for HECRecordStub and implement equals and hashCode methods for it

MoonBow-1 committed:
Rename HECTime.parsed() to HECTime.isParsed()

MoonBow-1 committed:
Rename JsonEvent timeNode1 field of HECTimeImpl to JsonEvent jsonEvent

MoonBow-1 committed:
Log an EventFieldException when parsing time in HECTimeImpl.instant

- Does not check that the JsonEvent has time before trying to parse it, since it catches the Exception anyway

eemhu commented on src/main/java/com/teragrep/cfe_16/HECBatch.java:
```diff
@@ -0,0 +1,127 @@
+        // Init the HECRecord as a Stub
+        HECRecord previousEvent = new HECRecordStub();
+
+        final JsonStreamParser parser = new JsonStreamParser(this.allEventInJSON);
```
Is it needed here to use both GSON and Jackson JSON parsers?

MoonBow-1 replied:
Gson's `JsonStreamParser` changed to Jackson's `JsonParser` in commit 58721402

- Also the dependency for Gson was removed, since it wasn't used anywhere else

eemhu commented on src/main/java/com/teragrep/cfe_16/HECBatch.java:
```diff
@@ -0,0 +1,127 @@
+             * Event field cannot be missing or blank. Throws an exception if this is the
+             * case.
+             */
+            final JsonEvent jsonEvent = new JsonEventImpl(new ObjectMapper().readTree(jsonObjectStr));
```
move `new ObjectMapper()` call to outside of loop

MoonBow-1 replied:
ObjectMapper is now a constructed field, done in commit 58721402

MoonBow-1 committed:
Combine HECTimeImpl's isParsed() and source() methods

- source() relies on the return value of isParsed(), since their values are connected

MoonBow-1 committed:
Make isStub a field for HECTimeStub and implement equals and hashCode methods for it

MoonBow-1 committed:
Make isStub a field for EventMessageStub and implement equals and hashCode methods for it

MoonBow-1 committed:
Simplify JsonEventImpl.hasTime() method

eemhu commented on src/test/java/com/teragrep/cfe_16/event/time/HECTimeImplTest.java:
```diff
@@ -0,0 +1,245 @@
+    @DisplayName("Time is reported and parsed when time is a double")
+    void timeIsReportedAndParsedWhenTimeIsADouble() {
+        final ObjectMapper objectMapper = new ObjectMapper();
+        ;
```
unnecessary ;

MoonBow-1 replied:
Removed in commit 95a636c0

eemhu commented on src/test/java/com/teragrep/cfe_16/event/time/HECTimeImplTest.java:
```diff
@@ -0,0 +1,245 @@
+    @DisplayName("Time is reported and parsed when time is exactly 13 digits")
+    void timeIsReportedAndParsedWhenTimeIsExactly13Digits() {
+        final ObjectMapper objectMapper = new ObjectMapper();
+        ;
```
;

MoonBow-1 replied:
Removed in commit 95a636c0

eemhu commented on src/test/java/com/teragrep/cfe_16/event/time/HECTimeImplWithFallbackTest.java:
```diff
@@ -0,0 +1,236 @@
+
+    @Test
+    @DisplayName(
+        "instant() returns the instant() value from the fallbackTime is currentTime.instant() is teh defaultValue"
```
typo teh->the

MoonBow-1 replied:
Fixed in commit 414233ce

eemhu commented on src/test/java/com/teragrep/cfe_16/it/ConfigurationIT.java:
```diff
@@ -60,38 +60,49 @@
+    public static void init() {
+        TestServerFactory serverFactory = new TestServerFactory();
+        try {
+            server = serverFactory.create(port, messageList, openCount, closeCount);
```
`Assertions.assertDoesNotThrow(() -> serverFactory.create(...));`

MoonBow-1 replied:
Try-catches and Exception throwing removed from tests in commit 72fbb4d8

MoonBow-1 committed:
Rename "SysLog" mentions to "Syslog"

MoonBow-1 committed:
Remove commented out code from HECServiceImpl

eemhu commented on src/test/java/com/teragrep/cfe_16/it/ConfigurationIT.java:
```diff
@@ -60,38 +60,49 @@
-        server.stop();
+    public static void close() {
+        try {
+            server.close();
```
also assertDoesNotThrow here

MoonBow-1 replied:
Try-catches and Exception throwing removed from tests in commit 72fbb4d8

eemhu commented on src/test/java/com/teragrep/cfe_16/it/ConfigurationIT.java:
```diff
@@ -101,5 +112,8 @@ public void instantiateConfigurationTest() {
         assertEquals(expected, configuration.toString());
+        assertEquals(0, messageList.size());
+        assertEquals(1, openCount.get());
+        assertEquals(0, closeCount.get()); // FIXME: does not close
```
Is the comment here still relevant?

MoonBow-1 replied:
Removed in commit e2ff0cc4

eemhu commented on src/test/java/com/teragrep/cfe_16/it/SendMultipleEventsIT.java:
```diff
@@ -79,73 +73,57 @@
+    @AfterAll
+    public static void close() {
+        try {
+            server.close();
```
assertDoesNotThrow

MoonBow-1 replied:
Try-catches and Exception throwing removed from tests in commit 72fbb4d8

eemhu commented on src/test/java/com/teragrep/cfe_16/it/SendSingleEventIT.java:
```diff
@@ -0,0 +1,142 @@
+    public static void init() {
+        TestServerFactory serverFactory = new TestServerFactory();
+        try {
+            server = serverFactory.create(SERVER_PORT, messageList, openCount, closeCount);
```
assert does not throw

MoonBow-1 replied:
Try-catches and Exception throwing removed from tests in commit 72fbb4d8

MoonBow-1 committed:
Remove Gson parser from HECBatch and use Jackson's parser instead

- Also removes the need for the Gson dependency, so that was removed from pom.xml

MoonBow-1 committed:
Fix type "teh" -> "the" in HECTimeImplWithFallbackTest

eemhu commented on src/test/java/com/teragrep/cfe_16/it/ServiceAndHECBatchIT.java:
```diff
@@ -0,0 +1,329 @@
+    public static void init() {
+        TestServerFactory serverFactory = new TestServerFactory();
+        try {
+            server = serverFactory.create(port, messageList, openCount, closeCount);
```
assertDoesNotThrow

MoonBow-1 replied:
Try-catches and Exception throwing removed from tests in commit 72fbb4d8

MoonBow-1 committed:
Remove try-catches and throws from tests

eemhu commented on src/test/java/com/teragrep/cfe_16/it/ServiceAndHECBatchIT.java:
```diff
@@ -0,0 +1,329 @@
+
+        ackRequestNode = objectMapper.createObjectNode();
+        try {
+            ackRequestNode = objectMapper.readTree(ackRequest);
```
assertDoesNotThrow

MoonBow-1 replied:
Try-catches and Exception throwing removed from tests in commit 72fbb4d8

MoonBow-1 committed:
Removed old FIXME from ConfigurationIT.instantiateConfigurationTest

eemhu commented on src/test/java/com/teragrep/cfe_16/server/TestServerFactory.java:
```diff
@@ -0,0 +1,105 @@
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.function.Supplier;
+
+public class TestServerFactory {
```
class & variables could be declared final.

MoonBow-1 replied:
Declared final in commit c17780f3

eemhu commented on src/test/java/com/teragrep/cfe_16/service/HECServiceImplTest.java:
```diff
@@ -0,0 +1,108 @@
+
+    @Test
+    @DisplayName("test the sendEvents does not throw an exception when JSON is malformed")
+    void testTheSendEventsDoesNotThrowAnExceptionWhenJsonIsMalformed() {
```
test case for when json is not malformed ?

MoonBow-1 replied:
Test case added in commit ce661472

MoonBow-1 committed:
Make class and ctor params final for TestServer

MoonBow-1 committed:
Add happy test for HECServiceImplTest and fix the previous test there

eemhu commented on src/test/java/com/teragrep/cfe_16/EventDataToSyslogMessageTest.java:
```diff
@@ -0,0 +1,462 @@
+        Set<SDElement> returnedSDElements1 = returnedMessage1.getSDElements();
+        Set<SDElement> expectedSDElements1 = expectedSyslogMessage1.getSDElements();
+
+        List<SDParam> supposedSDParams = new ArrayList<>();
```
perhaps the SDElements can be asserted like in nlf_01 https://github.com/teragrep/nlf_01/blob/e18425b3e0989cfee90e938827c3e5b8bdc2194c/src/test/java/com/teragrep/nlf_01/types/AppInsightTypeTest.java#L140-L142

MoonBow-1 replied:
New way of asserting SDElements done in commit 3b7752f5

MoonBow-1 committed:
Make SDElement tests more elegant in EventDataToSyslogMessageTest

- Renames test methods to be descriptive
- Uses the same way of collecting SDElements as NLF_01 does
- Removes star import from com.cloudbees.syslog package
- Makes variables final

MoonBow-1 committed:
Change HeaderInfo to use HttpServletRequest as its field

- Parses the header values from that

MoonBow-1 committed:
Change Boolean.hashCode(isStub()) to Boolean.hashCode(isStub)

MoonBow-1 committed:
Replace Boolean.hashCode() with Objects.hashCode()

eemhu commented on src/main/java/com/teragrep/cfe_16/bo/HeaderInfo.java:
```diff
@@ -46,66 +46,108 @@
-        this.xForwardedFor = xForwardedFor;
-        this.xForwardedHost = xForwardedHost;
-        this.xForwardedProto = xForwardedProto;
+    private final XForwardedForStub xForwardedForStub;
```
the stubs could be static also

MoonBow-1 replied:
Made Stubs static in commit f7868169

eemhu commented:
otherwise looks good

MoonBow-1 committed:
Make Stubs static for HeaderInfo

MoonBow-1 committed:
Remove jsonEvent.asPayloadJsonNode() from logged info in HECTimeImpl.instant

kortemik reviewed (commented):
looks good. has this been through jmeter?

kortemik reviewed (dismissed):
provided jmeter test plan gives

```
[ERROR]  2026-02-26 15:20:16  :  Technical error while processing request <[ExceptionEventContext{headerInfo=HeaderInfo{httpServletRequest=org.apache.catalina.connector.RequestFacade@448ed4ba}, userAgent='Apache-HttpClient/4.5.14 (Java/21.0.10)', uriPath='/services/collector', host='127.0.0.1'}]> correlation id <546e0517-b719-44ec-bd26-0a68cd2b6518>
com.teragrep.cfe_16.exceptionhandling.EventFieldException: Event field was not textual
        at com.teragrep.cfe_16.event.JsonEventImpl.asEventMessage(JsonEventImpl.java:89) ~[!/:?]
        at com.teragrep.cfe_16.HECBatch.toHECRecordList(HECBatch.java:127) ~[!/:?]
        at com.teragrep.cfe_16.service.HECServiceImpl.sendEvents(HECServiceImpl.java:183) ~[!/:?]
        at com.teragrep.cfe_16.rest.HECRestController.sendEvents(HECRestController.java:117) ~[!/:?]
        at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103) ~[?:?]
```

kortemik commented:
https://github.com/teragrep/cfe_16/issues/54

kortemik reviewed (commented):
seems like the exceptions were intended
